primary/src/block_synchronizer/handler.rs

```diff
@@ -0,0 +1,249 @@
+    certificate_deliver_timeout: Duration,
+}
+
+impl<PublicKey: VerifyingKey> BlockSynchronizerHandler<PublicKey> {
+    pub fn new(
+        tx_block_synchronizer: Sender<Command<PublicKey>>,
+        tx_core: Sender<PrimaryMessage<PublicKey>>,
+        certificate_store: Store<CertificateDigest, Certificate<PublicKey>>,
+        certificate_deliver_timeout: Duration,
+    ) -> Self {
+        Self {
+            tx_block_synchronizer,
+            tx_core,
+            certificate_store,
+            certificate_deliver_timeout,
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn wait_all(
+        &self,
+        certificates: Vec<Certificate<PublicKey>>,
+    ) -> Vec<Result<Certificate<PublicKey>, Error>> {
+        let futures: Vec<_> = certificates
+            .into_iter()
+            .map(|c| self.wait(c.digest()))
+            .collect();
+
+        join_all(futures).await
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn wait(&self, block_id: CertificateDigest) -> Result<Certificate<PublicKey>, Error> {
+        if let Ok(result) = timeout(
+            self.certificate_deliver_timeout,
+            self.certificate_store.notify_read(block_id),
+        )
+        .await
+        {
+            result
+                .map_err(|_| Error::Internal { block_id })?
+                .ok_or(Error::BlockNotFound { block_id })
+        } else {
+            Err(Error::BlockDeliveryTimeout { block_id })
+        }
+    }
+}
+
+#[async_trait]
+impl<PublicKey: VerifyingKey> Handler<PublicKey> for BlockSynchronizerHandler<PublicKey> {
+    /// The method will return a separate result for each requested block id.
+    /// If a certificate has been successfully retrieved (and processed via core
+    /// if has been fetched from peers) then an OK result will be returned with the
+    /// certificate value.
+    /// In case of error, the following outcomes are possible:
+    /// * BlockNotFound: Failed to retrieve the certificate either via the store or via the peers
+    /// * Internal: An internal error caused
+    /// * BlockDeliveryTimeout: Timed out while waiting for the certificate to become available
+    /// after submitting it for processing to core
+    #[instrument(level="debug", skip_all, fields(block_ids = ?block_ids))]
+    async fn get_and_synchronize_block_headers(
+        &self,
+        block_ids: Vec<CertificateDigest>,
+    ) -> Vec<Result<Certificate<PublicKey>, Error>> {
+        let sync_results = self.get_block_headers(block_ids).await;
+        let mut results: Vec<Result<Certificate<PublicKey>, Error>> = Vec::new();
+
+        // send certificates to core for processing and potential
+        // causal completion
+        let mut wait_for = Vec::new();
+
+        for result in sync_results {
+            match result {
+                Ok(block_header) => {
+                    // we need to perform causal completion since this
```

Comment:
Nit: I would place the comment a line below, to make sure the scoping helps the reader figure out what condition guarantees its validity.

Reply:
Good point!